// @ui8kit/aria/dist/accordion.cjs
'use strict';

// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/accordion/index.ts
function itemState(item, open) {
  const trigger = item.querySelector("[data-ui8kit-accordion-trigger]");
  const panel = item.querySelector("[data-ui8kit-accordion-content]");
  if (!trigger || !panel) {
    return;
  }
  item.setAttribute("data-state", open ? "open" : "closed");
  trigger.setAttribute("aria-expanded", open ? "true" : "false");
  if (open) {
    panel.removeAttribute("hidden");
  } else {
    panel.setAttribute("hidden", "hidden");
  }
}
function isMultiple(root) {
  return root.getAttribute("data-accordion-type") === "multiple";
}
function closeOthers(root, current) {
  const items = root.querySelectorAll("[data-accordion-item]");
  for (const item of items) {
    if (item !== current) {
      itemState(item, false);
    }
  }
}
function initAccordionRoot(root) {
  const items = root.querySelectorAll("[data-accordion-item]");
  for (const item of items) {
    const trigger = item.querySelector("[data-ui8kit-accordion-trigger]");
    const panel = item.querySelector("[data-ui8kit-accordion-content]");
    if (!trigger || !panel || item.dataset.ui8kitBound) {
      continue;
    }
    item.dataset.ui8kitBound = "1";
    trigger.type = "button";
    itemState(item, item.getAttribute("data-state") === "open");
    trigger.addEventListener("click", (event) => {
      event.preventDefault();
      const current = item;
      const next = item.getAttribute("data-state") !== "open";
      if (!isMultiple(root)) {
        closeOthers(root, current);
      }
      itemState(current, next);
    });
    trigger.addEventListener("keydown", (event) => {
      if (event.key === "Enter" || event.key === " ") {
        event.preventDefault();
        trigger.click();
      }
    });
  }
}
var accordion = {
  name: "accordion",
  init(root = document) {
    const roots = root.querySelectorAll('[data-ui8kit="accordion"]');
    for (const rootItem of roots) {
      initAccordionRoot(rootItem);
    }
    return () => {
    };
  }
};

// src/entries/accordion.ts
registerPattern(accordion);
var init2 = accordion.init;

exports.init = init2;
