// @ui8kit/aria/dist/all.iife.js
var ui8kit = (function (exports) {
  'use strict';

  // src/core/utils.ts
  function ready(fn) {
    if (typeof document === "undefined") {
      fn();
      return;
    }
    if (document.readyState === "loading") {
      document.addEventListener("DOMContentLoaded", fn, { once: true });
      return;
    }
    fn();
  }
  function byAttr(name, root = document) {
    return root.querySelectorAll(`[data-${name}]`);
  }
  var FOCUSABLE_SELECTOR = 'a[href], button, input, select, textarea, [tabindex]:not([tabindex="-1"])';
  function findFocusable(root) {
    const items = Array.from(root.querySelectorAll(FOCUSABLE_SELECTOR));
    return items.filter(
      (element) => !element.hasAttribute("disabled") && element.getAttribute("aria-hidden") !== "true"
    );
  }
  function isBrowser() {
    return typeof document !== "undefined" && typeof window !== "undefined";
  }

  // src/core/env.ts
  var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
  function isAutoInitEnabled() {
    return globalThis[AUTO_INIT_KEY] !== false;
  }
  function setAutoInitFlag(value) {
    globalThis[AUTO_INIT_KEY] = value;
  }

  // src/core/registry.ts
  var NAMESPACE_KEY = "ui8kit";
  var state = {
    patterns: /* @__PURE__ */ new Map(),
    scopes: /* @__PURE__ */ new Map()
  };
  var cachedNamespace = null;
  var coreHelpers = { ready, byAttr };
  var ready2 = ready;
  var byAttr2 = byAttr;
  function rootKey(root) {
    return root != null ? root : isBrowser() ? document : {};
  }
  function buildNamespace() {
    const existing = globalThis[NAMESPACE_KEY];
    if (existing) {
      return existing;
    }
    const ns = {
      core: coreHelpers,
      ready,
      byAttr,
      register: registerPattern,
      init,
      initPattern
    };
    if (isBrowser()) {
      globalThis[NAMESPACE_KEY] = ns;
    }
    return ns;
  }
  function getNamespace() {
    if (!cachedNamespace) {
      cachedNamespace = buildNamespace();
    }
    return cachedNamespace;
  }
  function registerPattern(pattern) {
    if (state.patterns.has(pattern.name)) {
      return;
    }
    state.patterns.set(pattern.name, pattern);
    getNamespace()[pattern.name] = pattern;
    if (isAutoInitEnabled() && isBrowser()) {
      ready(() => {
        initPattern(pattern.name);
      });
    }
  }
  function initPattern(name, root) {
    var _a;
    const pattern = state.patterns.get(name);
    if (!pattern) {
      return;
    }
    if (!isBrowser() && !root) {
      return;
    }
    const scope = rootKey(root);
    const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
    state.scopes.set(name, scopeMap);
    if (scopeMap.has(scope)) {
      return;
    }
    const result = pattern.init(scope);
    scopeMap.set(scope, typeof result === "function" ? result : noop);
  }
  function init(root) {
    for (const pattern of state.patterns.values()) {
      initPattern(pattern.name, root);
    }
  }
  function disposePattern(name, root) {
    const scope = rootKey(root);
    const scopeMap = state.scopes.get(name);
    if (!scopeMap) {
      return;
    }
    const dispose = scopeMap.get(scope);
    if (!dispose) {
      return;
    }
    try {
      dispose();
    } finally {
      scopeMap.delete(scope);
    }
  }
  function resetRegistry() {
    for (const [name, scopeMap] of state.scopes.entries()) {
    }
    state.patterns.clear();
    state.scopes.clear();
    if (isBrowser()) {
      delete globalThis[NAMESPACE_KEY];
    }
    cachedNamespace = null;
  }
  function noop() {
  }

  // src/patterns/dialog/index.ts
  var OPEN_ATTR = "data-state";
  var OPEN_VALUE = "open";
  var CLOSED_VALUE = "closed";
  var DIALOG_SELECTOR = '[data-ui8kit="dialog"], [data-ui8kit="sheet"], [data-ui8kit="alertdialog"]';
  var ACTIVE_DIALOG_SELECTOR = '[data-ui8kit="dialog"][data-state="open"], [data-ui8kit="sheet"][data-state="open"], [data-ui8kit="alertdialog"][data-state="open"]';
  var lastFocusByDialog = /* @__PURE__ */ new WeakMap();
  function toDialog(node) {
    var _a, _b;
    if (!node) {
      return null;
    }
    if (typeof node === "string") {
      return document.getElementById(node);
    }
    const element = node;
    if (typeof element.matches === "function" && element.matches(DIALOG_SELECTOR)) {
      return element;
    }
    const id = (_a = element.getAttribute) == null ? void 0 : _a.call(element, "data-ui8kit-dialog-target");
    if (id) {
      return document.getElementById(id);
    }
    return (_b = element.closest) == null ? void 0 : _b.call(element, DIALOG_SELECTOR);
  }
  function setOpen(dialog2, open2) {
    if (!dialog2) {
      return;
    }
    const overlay = dialog2.querySelector("[data-ui8kit-dialog-overlay]");
    if (open2) {
      dialog2.setAttribute(OPEN_ATTR, OPEN_VALUE);
      dialog2.removeAttribute("hidden");
      overlay == null ? void 0 : overlay.removeAttribute("hidden");
      const firstFocusable = findFocusable(dialog2)[0];
      firstFocusable == null ? void 0 : firstFocusable.focus();
      dialog2.dataset.trapped = "1";
    } else {
      dialog2.setAttribute(OPEN_ATTR, CLOSED_VALUE);
      dialog2.setAttribute("hidden", "hidden");
      overlay == null ? void 0 : overlay.setAttribute("hidden", "hidden");
      const last = lastFocusByDialog.get(dialog2);
      if (last && document.contains(last)) {
        last.focus();
      }
      lastFocusByDialog.delete(dialog2);
      delete dialog2.dataset.trapped;
    }
  }
  function open(dialogRoot) {
    const dialog2 = toDialog(dialogRoot);
    if (!dialog2) {
      return;
    }
    const sourceElement = dialogRoot && dialogRoot instanceof HTMLElement && dialogRoot !== dialog2 ? dialogRoot : document.activeElement;
    if (sourceElement) {
      lastFocusByDialog.set(dialog2, sourceElement);
    }
    setOpen(dialog2, true);
  }
  function close(dialogRoot) {
    setOpen(toDialog(dialogRoot), false);
  }
  function handleDocumentKeydown(event) {
    if (event.key === "Escape") {
      const dialog2 = document.querySelector(ACTIVE_DIALOG_SELECTOR);
      if (dialog2) {
        close(dialog2);
        event.preventDefault();
      }
      return;
    }
    if (event.key !== "Tab") {
      return;
    }
    const activeDialog = document.querySelector(ACTIVE_DIALOG_SELECTOR);
    if (!activeDialog || !activeDialog.contains(event.target)) {
      return;
    }
    const items = findFocusable(activeDialog);
    if (items.length === 0) {
      event.preventDefault();
      return;
    }
    const first = items[0];
    const last = items[items.length - 1];
    if (!first || !last) {
      event.preventDefault();
      return;
    }
    const current = document.activeElement;
    if (event.shiftKey) {
      if (current === first || current === activeDialog) {
        last.focus();
        event.preventDefault();
      }
    } else if (current === last || current === activeDialog) {
      first.focus();
      event.preventDefault();
    }
  }
  function handleDocumentClick(event) {
    const target = event.target;
    if (!target) {
      return;
    }
    const openButton = target.closest("[data-ui8kit-dialog-open]");
    if (openButton) {
      open(openButton);
      event.preventDefault();
      return;
    }
    if (target.closest("[data-ui8kit-dialog-close]")) {
      close(target.closest(DIALOG_SELECTOR));
      event.preventDefault();
      return;
    }
    if (target.closest("[data-ui8kit-dialog-overlay]")) {
      close(target.closest(DIALOG_SELECTOR));
    }
  }
  var dialog = {
    name: "dialog",
    init(root = document) {
      const scope = root;
      const dialogs = scope.querySelectorAll(DIALOG_SELECTOR);
      for (const node of dialogs) {
        setOpen(node, node.getAttribute(OPEN_ATTR) === OPEN_VALUE);
      }
      document.addEventListener("click", handleDocumentClick);
      document.addEventListener("keydown", handleDocumentKeydown);
      return () => {
        document.removeEventListener("click", handleDocumentClick);
        document.removeEventListener("keydown", handleDocumentKeydown);
      };
    },
    open,
    close
  };

  // src/patterns/accordion/index.ts
  function itemState(item, open2) {
    const trigger = item.querySelector("[data-ui8kit-accordion-trigger]");
    const panel = item.querySelector("[data-ui8kit-accordion-content]");
    if (!trigger || !panel) {
      return;
    }
    item.setAttribute("data-state", open2 ? "open" : "closed");
    trigger.setAttribute("aria-expanded", open2 ? "true" : "false");
    if (open2) {
      panel.removeAttribute("hidden");
    } else {
      panel.setAttribute("hidden", "hidden");
    }
  }
  function isMultiple(root) {
    return root.getAttribute("data-accordion-type") === "multiple";
  }
  function closeOthers(root, current) {
    const items = root.querySelectorAll("[data-accordion-item]");
    for (const item of items) {
      if (item !== current) {
        itemState(item, false);
      }
    }
  }
  function initAccordionRoot(root) {
    const items = root.querySelectorAll("[data-accordion-item]");
    for (const item of items) {
      const trigger = item.querySelector("[data-ui8kit-accordion-trigger]");
      const panel = item.querySelector("[data-ui8kit-accordion-content]");
      if (!trigger || !panel || item.dataset.ui8kitBound) {
        continue;
      }
      item.dataset.ui8kitBound = "1";
      trigger.type = "button";
      itemState(item, item.getAttribute("data-state") === "open");
      trigger.addEventListener("click", (event) => {
        event.preventDefault();
        const current = item;
        const next = item.getAttribute("data-state") !== "open";
        if (!isMultiple(root)) {
          closeOthers(root, current);
        }
        itemState(current, next);
      });
      trigger.addEventListener("keydown", (event) => {
        if (event.key === "Enter" || event.key === " ") {
          event.preventDefault();
          trigger.click();
        }
      });
    }
  }
  var accordion = {
    name: "accordion",
    init(root = document) {
      const roots = root.querySelectorAll('[data-ui8kit="accordion"]');
      for (const rootItem of roots) {
        initAccordionRoot(rootItem);
      }
      return () => {
      };
    }
  };

  // src/patterns/tabs/index.ts
  function activate(root, value, useFocus) {
    const triggers = root.querySelectorAll("[data-tabs-trigger]");
    const panels = root.querySelectorAll("[data-tabs-panel]");
    for (const trigger of triggers) {
      const selected = trigger.getAttribute("data-tabs-value") === value;
      trigger.setAttribute("aria-selected", selected ? "true" : "false");
      trigger.tabIndex = selected ? 0 : -1;
      if (selected && useFocus) {
        trigger.focus();
      }
    }
    for (const panel of panels) {
      panel.hidden = panel.getAttribute("data-tabs-value") !== value;
    }
  }
  function defaultValue(root) {
    var _a;
    const rootValue = root.getAttribute("data-tabs-value");
    if (rootValue) return rootValue;
    const selected = root.querySelector('[data-tabs-trigger][aria-selected="true"]');
    if (selected == null ? void 0 : selected.getAttribute("data-tabs-value")) {
      return selected.getAttribute("data-tabs-value");
    }
    return ((_a = root.querySelector("[data-tabs-trigger]")) == null ? void 0 : _a.getAttribute("data-tabs-value")) || "";
  }
  function onKeydown(root, event) {
    const triggers = Array.from(root.querySelectorAll("[data-tabs-trigger]"));
    if (!triggers.length) return;
    const trigger = event.target.closest("[data-tabs-trigger]");
    if (!trigger) return;
    const index = triggers.indexOf(trigger);
    if (index < 0) return;
    if (event.key === "ArrowRight" || event.key === "ArrowDown") {
      const next = triggers[(index + 1) % triggers.length];
      event.preventDefault();
      activate(root, next.getAttribute("data-tabs-value") || "", true);
    } else if (event.key === "ArrowLeft" || event.key === "ArrowUp") {
      const prev = triggers[(index - 1 + triggers.length) % triggers.length];
      event.preventDefault();
      activate(root, prev.getAttribute("data-tabs-value") || "", true);
    }
  }
  var tabs = {
    name: "tabs",
    init(root = document) {
      const roots = root.querySelectorAll('[data-ui8kit="tabs"]');
      for (const tabsRoot of roots) {
        if (tabsRoot.dataset.ui8kitBound) continue;
        tabsRoot.dataset.ui8kitBound = "1";
        const value = defaultValue(tabsRoot);
        if (value) {
          activate(tabsRoot, value, false);
        }
        tabsRoot.addEventListener("click", (event) => {
          const trigger = event.target.closest("[data-tabs-trigger]");
          if (!trigger) return;
          const value2 = trigger.getAttribute("data-tabs-value");
          if (!value2) return;
          event.preventDefault();
          activate(tabsRoot, value2, false);
        });
        tabsRoot.addEventListener("keydown", (event) => {
          onKeydown(tabsRoot, event);
        });
      }
      return () => {
      };
    }
  };

  // src/patterns/combobox/index.ts
  var COMBOBOX_SELECTOR = '[data-ui8kit="combobox"]';
  function setOpen2(root, state2) {
    const list = root.querySelector('[role="listbox"], ul');
    const trigger = root.querySelector("[data-combobox-toggle]");
    const input = root.querySelector("input");
    if (!list || !input) {
      return;
    }
    root.dataset.state = state2 ? "open" : "closed";
    trigger == null ? void 0 : trigger.setAttribute("aria-expanded", String(state2));
    input.setAttribute("aria-expanded", String(state2));
    if (state2) {
      list.removeAttribute("hidden");
    } else {
      list.setAttribute("hidden", "hidden");
    }
  }
  function getVisibleOptions(root) {
    const options = Array.from(root.querySelectorAll("[data-combobox-option]"));
    return options.filter((opt) => {
      return opt.style.display !== "none" && opt.getAttribute("aria-disabled") !== "true";
    });
  }
  function syncVisibleSelection(root, option) {
    const options = root.querySelectorAll("[data-combobox-option]");
    for (const item of options) {
      item.classList.remove("ui-combobox-option-active");
      item.setAttribute("aria-selected", "false");
    }
    if (option) {
      option.classList.add("ui-combobox-option-active");
      option.setAttribute("aria-selected", "true");
    }
  }
  function filterOptions(root) {
    const input = root.querySelector("input");
    if (!input) return;
    const options = root.querySelectorAll("[data-combobox-option]");
    const phrase = input.value.trim().toLowerCase();
    for (const option of options) {
      const text = (option.textContent || "").toLowerCase();
      const visible = phrase.length === 0 || text.includes(phrase);
      option.style.display = visible ? "" : "none";
    }
  }
  function selectOption(root, option) {
    if (!option) return;
    const input = root.querySelector("input");
    if (!input) return;
    const value = option.getAttribute("data-combobox-value") || option.textContent || "";
    input.value = value;
    setOpen2(root, false);
  }
  function bindCombobox(cb) {
    const input = cb.querySelector("input");
    const list = cb.querySelector('[role="listbox"], ul');
    const toggle3 = cb.querySelector("[data-combobox-toggle]");
    if (!input || !list) {
      return () => {
      };
    }
    const onFocus = () => setOpen2(cb, true);
    const onInput = () => {
      setOpen2(cb, true);
      filterOptions(cb);
    };
    const onKeydown2 = (event) => {
      const visible = getVisibleOptions(cb);
      if (event.key === "ArrowDown" || event.key === "ArrowUp") {
        if (visible.length === 0) return;
        const active = cb.querySelector(".ui-combobox-option-active");
        const index = active ? visible.indexOf(active) : -1;
        const next = event.key === "ArrowDown" ? (index + 1) % visible.length : (index - 1 + visible.length) % visible.length;
        syncVisibleSelection(cb, visible[next]);
        const target = visible[next];
        if (target && typeof target.scrollIntoView === "function") {
          try {
            target.scrollIntoView({ block: "nearest" });
          } catch (e) {
          }
        }
        event.preventDefault();
        return;
      }
      if (event.key === "Enter") {
        selectOption(cb, cb.querySelector(".ui-combobox-option-active"));
        event.preventDefault();
        return;
      }
      if (event.key === "Escape") {
        setOpen2(cb, false);
        event.preventDefault();
      }
    };
    const onToggleClick = () => {
      const isOpen = cb.dataset.state === "open";
      setOpen2(cb, !isOpen);
    };
    const optionMousedown = (event) => event.preventDefault();
    const optionClicks = [];
    for (const option of cb.querySelectorAll("[data-combobox-option]")) {
      const handler = () => selectOption(cb, option);
      option.addEventListener("mousedown", optionMousedown);
      option.addEventListener("click", handler);
      optionClicks.push({ el: option, handler });
    }
    input.addEventListener("focus", onFocus);
    input.addEventListener("input", onInput);
    input.addEventListener("keydown", onKeydown2);
    toggle3 == null ? void 0 : toggle3.addEventListener("click", onToggleClick);
    filterOptions(cb);
    return () => {
      input.removeEventListener("focus", onFocus);
      input.removeEventListener("input", onInput);
      input.removeEventListener("keydown", onKeydown2);
      toggle3 == null ? void 0 : toggle3.removeEventListener("click", onToggleClick);
      for (const { el, handler } of optionClicks) {
        el.removeEventListener("mousedown", optionMousedown);
        el.removeEventListener("click", handler);
      }
    };
  }
  function handleOutsideClick(event) {
    var _a;
    const target = event.target;
    if (!target || ((_a = target.closest) == null ? void 0 : _a.call(target, COMBOBOX_SELECTOR))) {
      return;
    }
    for (const item of document.querySelectorAll(COMBOBOX_SELECTOR)) {
      setOpen2(item, false);
    }
  }
  var combobox = {
    name: "combobox",
    init(root = document) {
      const scope = root;
      const roots = scope.querySelectorAll(COMBOBOX_SELECTOR);
      const disposers = [];
      for (const cb of roots) {
        if (cb.dataset.ui8kitBound) continue;
        cb.dataset.ui8kitBound = "1";
        disposers.push(bindCombobox(cb));
      }
      document.addEventListener("click", handleOutsideClick);
      return () => {
        document.removeEventListener("click", handleOutsideClick);
        for (const dispose of disposers) {
          dispose();
        }
      };
    }
  };

  // src/patterns/tooltip/index.ts
  function show(root) {
    const content = root.querySelector('[role="tooltip"]');
    if (!content) {
      return;
    }
    content.removeAttribute("hidden");
    root.setAttribute("data-state", "open");
    content.setAttribute("aria-hidden", "false");
  }
  function hide(root) {
    const content = root.querySelector('[role="tooltip"]');
    if (!content) {
      return;
    }
    content.setAttribute("hidden", "hidden");
    root.setAttribute("data-state", "closed");
    content.setAttribute("aria-hidden", "true");
  }
  var tooltip = {
    name: "tooltip",
    init(root = document) {
      const tooltips = root.querySelectorAll('[data-ui8kit="tooltip"]');
      for (const tooltipRoot of tooltips) {
        if (tooltipRoot.dataset.ui8kitBound) continue;
        tooltipRoot.addEventListener("mouseenter", () => show(tooltipRoot));
        tooltipRoot.addEventListener("focusin", () => show(tooltipRoot));
        tooltipRoot.addEventListener("mouseleave", () => hide(tooltipRoot));
        tooltipRoot.addEventListener("focusout", () => hide(tooltipRoot));
        tooltipRoot.dataset.ui8kitBound = "1";
      }
      return () => {
      };
    }
  };

  // src/patterns/alert/index.ts
  var alertPattern = {
    name: "alert",
    init(root = document) {
      const nodes = root.querySelectorAll('[data-ui8kit="alert"], [role="alert"]');
      for (const alert of nodes) {
        if (!alert.hasAttribute("role")) {
          alert.setAttribute("role", "status");
        }
        if (!alert.hasAttribute("aria-live")) {
          alert.setAttribute("aria-live", "polite");
        }
        if (!alert.hasAttribute("aria-atomic")) {
          alert.setAttribute("aria-atomic", "true");
        }
      }
      return () => {
      };
    }
  };

  // src/patterns/disclosure/index.ts
  function setState(root, open2) {
    const trigger = root.querySelector("[data-disclosure-trigger]");
    const content = root.querySelector("[data-disclosure-content]");
    if (!trigger || !content) {
      return;
    }
    trigger.setAttribute("aria-expanded", String(open2));
    if (open2) {
      content.removeAttribute("hidden");
    } else {
      content.setAttribute("hidden", "hidden");
    }
  }
  var disclosure = {
    name: "disclosure",
    init(root = document) {
      const roots = root.querySelectorAll('[data-ui8kit="disclosure"]');
      for (const disclosureRoot of roots) {
        const trigger = disclosureRoot.querySelector("[data-disclosure-trigger]");
        const content = disclosureRoot.querySelector("[data-disclosure-content]");
        if (!trigger || !content || disclosureRoot.dataset.ui8kitBound) continue;
        const initial = trigger.getAttribute("aria-expanded") === "true";
        setState(disclosureRoot, initial);
        trigger.addEventListener("click", () => {
          const next = trigger.getAttribute("aria-expanded") !== "true";
          setState(disclosureRoot, next);
        });
        trigger.addEventListener("keydown", (event) => {
          if (event.key === "Enter" || event.key === " ") {
            event.preventDefault();
            trigger.click();
          }
        });
        disclosureRoot.dataset.ui8kitBound = "1";
      }
      return () => {
      };
    }
  };

  // src/patterns/menu/index.ts
  function makeItemsFocusable(items) {
    for (const [index, item] of items.entries()) {
      item.tabIndex = index === 0 ? 0 : -1;
    }
  }
  function moveFocus(items, current, delta) {
    const next = (current + delta + items.length) % items.length;
    items[current].tabIndex = -1;
    items[next].tabIndex = 0;
    items[next].focus();
    return next;
  }
  var menu = {
    name: "menu",
    init(root = document) {
      const menus = root.querySelectorAll('[data-ui8kit="menu"]');
      for (const menuRoot of menus) {
        const items = Array.from(menuRoot.querySelectorAll("[data-menu-item]"));
        if (!items.length || menuRoot.dataset.ui8kitBound) continue;
        makeItemsFocusable(items);
        let current = 0;
        for (const item of items) {
          item.addEventListener("keydown", (event) => {
            if (event.key === "ArrowDown") {
              current = moveFocus(items, current, 1);
              event.preventDefault();
            }
            if (event.key === "ArrowUp") {
              current = moveFocus(items, current, -1);
              event.preventDefault();
            }
            if (event.key === "Home") {
              current = moveFocus(items, current, -current);
              event.preventDefault();
            }
            if (event.key === "End") {
              current = moveFocus(items, current, items.length - 1 - current);
              event.preventDefault();
            }
            if (event.key === "Escape") {
              menuRoot.setAttribute("hidden", "hidden");
              const button = document.querySelector(`[data-menubutton-target="${menuRoot.id}"]`);
              button == null ? void 0 : button.focus();
            }
          });
          item.addEventListener("click", () => {
            item.focus();
          });
        }
        menuRoot.dataset.ui8kitBound = "1";
      }
      return () => {
      };
    }
  };

  // src/patterns/menubutton/index.ts
  function getMenu(button) {
    const id = button.dataset.menubuttonTarget || button.getAttribute("aria-controls");
    if (!id) return null;
    const menu2 = document.getElementById(id);
    return menu2 && menu2.matches('[data-ui8kit="menu"]') ? menu2 : null;
  }
  function closeAllMenus() {
    const menus = document.querySelectorAll('[data-ui8kit="menu"]');
    for (const menu2 of menus) {
      menu2.setAttribute("hidden", "hidden");
      menu2.dataset.state = "closed";
    }
    const buttons = document.querySelectorAll('[data-ui8kit="menubutton"]');
    for (const btn of buttons) {
      btn.setAttribute("aria-expanded", "false");
    }
  }
  function toggle(button) {
    const menu2 = getMenu(button);
    if (!menu2) return;
    const isOpen = menu2.dataset.state === "open";
    if (isOpen) {
      menu2.setAttribute("hidden", "hidden");
      menu2.dataset.state = "closed";
      button.setAttribute("aria-expanded", "false");
    } else {
      closeAllMenus();
      menu2.removeAttribute("hidden");
      menu2.dataset.state = "open";
      button.setAttribute("aria-expanded", "true");
      const first = menu2.querySelector("[data-menu-item]");
      first == null ? void 0 : first.focus();
    }
  }
  function handleDocumentClick2(event) {
    const target = event.target;
    if (!target) return;
    if (!target.closest('[data-ui8kit="menubutton"], [data-ui8kit="menu"]')) {
      closeAllMenus();
    }
  }
  var menubutton = {
    name: "menubutton",
    init(root = document) {
      const scope = root;
      const buttons = scope.querySelectorAll('[data-ui8kit="menubutton"]');
      const cleanups = [];
      for (const button of buttons) {
        if (button.dataset.ui8kitBound) continue;
        button.setAttribute("aria-expanded", "false");
        const onClick = (event) => {
          event.preventDefault();
          toggle(button);
        };
        const onKeydown2 = (event) => {
          if (event.key === "ArrowDown" || event.key === "Enter" || event.key === " ") {
            event.preventDefault();
            toggle(button);
          }
        };
        button.addEventListener("click", onClick);
        button.addEventListener("keydown", onKeydown2);
        button.dataset.ui8kitBound = "1";
        cleanups.push(() => {
          button.removeEventListener("click", onClick);
          button.removeEventListener("keydown", onKeydown2);
        });
      }
      document.addEventListener("click", handleDocumentClick2);
      return () => {
        document.removeEventListener("click", handleDocumentClick2);
        for (const fn of cleanups) fn();
      };
    },
    closeAll: closeAllMenus
  };

  // src/patterns/listbox/index.ts
  function setSelected(root, target) {
    const options = root.querySelectorAll('[role="option"]');
    for (const option of options) {
      option.classList.remove("ui-listbox-option-active");
      option.setAttribute("aria-selected", "false");
      option.setAttribute("tabindex", "-1");
    }
    target.classList.add("ui-listbox-option-active");
    target.setAttribute("aria-selected", "true");
    target.setAttribute("tabindex", "0");
    target.focus();
  }
  var listbox = {
    name: "listbox",
    init(root = document) {
      const lists = root.querySelectorAll('[data-ui8kit="listbox"]');
      for (const list of lists) {
        const options = Array.from(list.querySelectorAll('[role="option"]'));
        if (!options.length || list.dataset.ui8kitBound) continue;
        for (let i = 0; i < options.length; i++) {
          options[i].tabIndex = i === 0 ? 0 : -1;
          options[i].addEventListener("click", () => {
            setSelected(list, options[i]);
          });
        }
        list.addEventListener("keydown", (event) => {
          if (event.key === "Home") {
            setSelected(list, options[0]);
            event.preventDefault();
          }
          if (event.key === "End") {
            setSelected(list, options[options.length - 1]);
            event.preventDefault();
          }
          if (event.key === "ArrowDown") {
            const current = options.indexOf(document.activeElement);
            const next = (current + 1) % options.length;
            setSelected(list, options[next]);
            event.preventDefault();
          }
          if (event.key === "ArrowUp") {
            const current = options.indexOf(document.activeElement);
            const prev = (current - 1 + options.length) % options.length;
            setSelected(list, options[prev]);
            event.preventDefault();
          }
        });
        list.dataset.ui8kitBound = "1";
      }
      return () => {
      };
    }
  };

  // src/patterns/switch/index.ts
  function toggle2(control) {
    const checked = control.getAttribute("aria-checked") === "true";
    const next = !checked;
    control.setAttribute("aria-checked", String(next));
    control.setAttribute("data-state", next ? "checked" : "unchecked");
    control.dataset.value = String(next);
  }
  var switchPattern = {
    name: "switch",
    init(root = document) {
      var _a;
      const switches = root.querySelectorAll('[data-ui8kit="switch"], [role="switch"]');
      for (const item of switches) {
        if (item.dataset.ui8kitBound) continue;
        item.dataset.ui8kitBound = "1";
        const control = item.querySelector("[data-switch-control]") || item;
        control.setAttribute("role", "switch");
        control.setAttribute("tabindex", (_a = control.getAttribute("tabindex")) != null ? _a : "0");
        if (!control.hasAttribute("aria-checked")) {
          control.setAttribute("aria-checked", "false");
        }
        const handler = (event) => {
          event.preventDefault();
          toggle2(control);
        };
        control.addEventListener("click", handler);
        control.addEventListener("keydown", (event) => {
          if (event.key === " " || event.key === "Enter") {
            handler(event);
          }
        });
      }
      return () => {
      };
    }
  };

  // src/entries/all.ts
  registerPattern(dialog);
  registerPattern(accordion);
  registerPattern(tabs);
  registerPattern(combobox);
  registerPattern(tooltip);
  registerPattern(alertPattern);
  registerPattern(disclosure);
  registerPattern(menu);
  registerPattern(menubutton);
  registerPattern(listbox);
  registerPattern(switchPattern);
  var ui8kit = getNamespace();

  exports.AUTO_INIT_KEY = AUTO_INIT_KEY;
  exports.accordion = accordion;
  exports.alertPattern = alertPattern;
  exports.byAttr = byAttr2;
  exports.combobox = combobox;
  exports.coreHelpers = coreHelpers;
  exports.dialog = dialog;
  exports.disclosure = disclosure;
  exports.disposePattern = disposePattern;
  exports.findFocusable = findFocusable;
  exports.getNamespace = getNamespace;
  exports.init = init;
  exports.initPattern = initPattern;
  exports.isAutoInitEnabled = isAutoInitEnabled;
  exports.isBrowser = isBrowser;
  exports.listbox = listbox;
  exports.menu = menu;
  exports.menubutton = menubutton;
  exports.ready = ready2;
  exports.registerPattern = registerPattern;
  exports.resetRegistry = resetRegistry;
  exports.setAutoInitFlag = setAutoInitFlag;
  exports.switchPattern = switchPattern;
  exports.tabs = tabs;
  exports.tooltip = tooltip;
  exports.ui8kit = ui8kit;

  return exports;

})({});
