// @ui8kit/aria/dist/combobox.mjs
// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/combobox/index.ts
var COMBOBOX_SELECTOR = '[data-ui8kit="combobox"]';
function setOpen(root, state2) {
  const list = root.querySelector('[role="listbox"], ul');
  const trigger = root.querySelector("[data-combobox-toggle]");
  const input = root.querySelector("input");
  if (!list || !input) {
    return;
  }
  root.dataset.state = state2 ? "open" : "closed";
  trigger == null ? void 0 : trigger.setAttribute("aria-expanded", String(state2));
  input.setAttribute("aria-expanded", String(state2));
  if (state2) {
    list.removeAttribute("hidden");
  } else {
    list.setAttribute("hidden", "hidden");
  }
}
function getVisibleOptions(root) {
  const options = Array.from(root.querySelectorAll("[data-combobox-option]"));
  return options.filter((opt) => {
    return opt.style.display !== "none" && opt.getAttribute("aria-disabled") !== "true";
  });
}
function syncVisibleSelection(root, option) {
  const options = root.querySelectorAll("[data-combobox-option]");
  for (const item of options) {
    item.classList.remove("ui-combobox-option-active");
    item.setAttribute("aria-selected", "false");
  }
  if (option) {
    option.classList.add("ui-combobox-option-active");
    option.setAttribute("aria-selected", "true");
  }
}
function filterOptions(root) {
  const input = root.querySelector("input");
  if (!input) return;
  const options = root.querySelectorAll("[data-combobox-option]");
  const phrase = input.value.trim().toLowerCase();
  for (const option of options) {
    const text = (option.textContent || "").toLowerCase();
    const visible = phrase.length === 0 || text.includes(phrase);
    option.style.display = visible ? "" : "none";
  }
}
function selectOption(root, option) {
  if (!option) return;
  const input = root.querySelector("input");
  if (!input) return;
  const value = option.getAttribute("data-combobox-value") || option.textContent || "";
  input.value = value;
  setOpen(root, false);
}
function bindCombobox(cb) {
  const input = cb.querySelector("input");
  const list = cb.querySelector('[role="listbox"], ul');
  const toggle = cb.querySelector("[data-combobox-toggle]");
  if (!input || !list) {
    return () => {
    };
  }
  const onFocus = () => setOpen(cb, true);
  const onInput = () => {
    setOpen(cb, true);
    filterOptions(cb);
  };
  const onKeydown = (event) => {
    const visible = getVisibleOptions(cb);
    if (event.key === "ArrowDown" || event.key === "ArrowUp") {
      if (visible.length === 0) return;
      const active = cb.querySelector(".ui-combobox-option-active");
      const index = active ? visible.indexOf(active) : -1;
      const next = event.key === "ArrowDown" ? (index + 1) % visible.length : (index - 1 + visible.length) % visible.length;
      syncVisibleSelection(cb, visible[next]);
      const target = visible[next];
      if (target && typeof target.scrollIntoView === "function") {
        try {
          target.scrollIntoView({ block: "nearest" });
        } catch (e) {
        }
      }
      event.preventDefault();
      return;
    }
    if (event.key === "Enter") {
      selectOption(cb, cb.querySelector(".ui-combobox-option-active"));
      event.preventDefault();
      return;
    }
    if (event.key === "Escape") {
      setOpen(cb, false);
      event.preventDefault();
    }
  };
  const onToggleClick = () => {
    const isOpen = cb.dataset.state === "open";
    setOpen(cb, !isOpen);
  };
  const optionMousedown = (event) => event.preventDefault();
  const optionClicks = [];
  for (const option of cb.querySelectorAll("[data-combobox-option]")) {
    const handler = () => selectOption(cb, option);
    option.addEventListener("mousedown", optionMousedown);
    option.addEventListener("click", handler);
    optionClicks.push({ el: option, handler });
  }
  input.addEventListener("focus", onFocus);
  input.addEventListener("input", onInput);
  input.addEventListener("keydown", onKeydown);
  toggle == null ? void 0 : toggle.addEventListener("click", onToggleClick);
  filterOptions(cb);
  return () => {
    input.removeEventListener("focus", onFocus);
    input.removeEventListener("input", onInput);
    input.removeEventListener("keydown", onKeydown);
    toggle == null ? void 0 : toggle.removeEventListener("click", onToggleClick);
    for (const { el, handler } of optionClicks) {
      el.removeEventListener("mousedown", optionMousedown);
      el.removeEventListener("click", handler);
    }
  };
}
function handleOutsideClick(event) {
  var _a;
  const target = event.target;
  if (!target || ((_a = target.closest) == null ? void 0 : _a.call(target, COMBOBOX_SELECTOR))) {
    return;
  }
  for (const item of document.querySelectorAll(COMBOBOX_SELECTOR)) {
    setOpen(item, false);
  }
}
var combobox = {
  name: "combobox",
  init(root = document) {
    const scope = root;
    const roots = scope.querySelectorAll(COMBOBOX_SELECTOR);
    const disposers = [];
    for (const cb of roots) {
      if (cb.dataset.ui8kitBound) continue;
      cb.dataset.ui8kitBound = "1";
      disposers.push(bindCombobox(cb));
    }
    document.addEventListener("click", handleOutsideClick);
    return () => {
      document.removeEventListener("click", handleOutsideClick);
      for (const dispose of disposers) {
        dispose();
      }
    };
  }
};

// src/entries/combobox.ts
registerPattern(combobox);
var init2 = combobox.init;

export { init2 as init };
