// @ui8kit/aria/dist/core.cjs
'use strict';

// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}
function setAutoInitFlag(value) {
  globalThis[AUTO_INIT_KEY] = value;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
var ready2 = ready;
var byAttr2 = byAttr;
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function disposePattern(name, root) {
  const scope = rootKey(root);
  const scopeMap = state.scopes.get(name);
  if (!scopeMap) {
    return;
  }
  const dispose = scopeMap.get(scope);
  if (!dispose) {
    return;
  }
  try {
    dispose();
  } finally {
    scopeMap.delete(scope);
  }
}
function resetRegistry() {
  for (const [name, scopeMap] of state.scopes.entries()) {
  }
  state.patterns.clear();
  state.scopes.clear();
  if (isBrowser()) {
    delete globalThis[NAMESPACE_KEY];
  }
  cachedNamespace = null;
}
function noop() {
}

exports.AUTO_INIT_KEY = AUTO_INIT_KEY;
exports.byAttr = byAttr2;
exports.coreHelpers = coreHelpers;
exports.disposePattern = disposePattern;
exports.getNamespace = getNamespace;
exports.init = init;
exports.initPattern = initPattern;
exports.isAutoInitEnabled = isAutoInitEnabled;
exports.ready = ready2;
exports.registerPattern = registerPattern;
exports.resetRegistry = resetRegistry;
exports.setAutoInitFlag = setAutoInitFlag;
