// @ui8kit/aria/dist/dialog.cjs
'use strict';

// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
var FOCUSABLE_SELECTOR = 'a[href], button, input, select, textarea, [tabindex]:not([tabindex="-1"])';
function findFocusable(root) {
  const items = Array.from(root.querySelectorAll(FOCUSABLE_SELECTOR));
  return items.filter(
    (element) => !element.hasAttribute("disabled") && element.getAttribute("aria-hidden") !== "true"
  );
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/dialog/index.ts
var OPEN_ATTR = "data-state";
var OPEN_VALUE = "open";
var CLOSED_VALUE = "closed";
var DIALOG_SELECTOR = '[data-ui8kit="dialog"], [data-ui8kit="sheet"], [data-ui8kit="alertdialog"]';
var ACTIVE_DIALOG_SELECTOR = '[data-ui8kit="dialog"][data-state="open"], [data-ui8kit="sheet"][data-state="open"], [data-ui8kit="alertdialog"][data-state="open"]';
var lastFocusByDialog = /* @__PURE__ */ new WeakMap();
function toDialog(node) {
  var _a, _b;
  if (!node) {
    return null;
  }
  if (typeof node === "string") {
    return document.getElementById(node);
  }
  const element = node;
  if (typeof element.matches === "function" && element.matches(DIALOG_SELECTOR)) {
    return element;
  }
  const id = (_a = element.getAttribute) == null ? void 0 : _a.call(element, "data-ui8kit-dialog-target");
  if (id) {
    return document.getElementById(id);
  }
  return (_b = element.closest) == null ? void 0 : _b.call(element, DIALOG_SELECTOR);
}
function setOpen(dialog2, open3) {
  if (!dialog2) {
    return;
  }
  const overlay = dialog2.querySelector("[data-ui8kit-dialog-overlay]");
  if (open3) {
    dialog2.setAttribute(OPEN_ATTR, OPEN_VALUE);
    dialog2.removeAttribute("hidden");
    overlay == null ? void 0 : overlay.removeAttribute("hidden");
    const firstFocusable = findFocusable(dialog2)[0];
    firstFocusable == null ? void 0 : firstFocusable.focus();
    dialog2.dataset.trapped = "1";
  } else {
    dialog2.setAttribute(OPEN_ATTR, CLOSED_VALUE);
    dialog2.setAttribute("hidden", "hidden");
    overlay == null ? void 0 : overlay.setAttribute("hidden", "hidden");
    const last = lastFocusByDialog.get(dialog2);
    if (last && document.contains(last)) {
      last.focus();
    }
    lastFocusByDialog.delete(dialog2);
    delete dialog2.dataset.trapped;
  }
}
function open(dialogRoot) {
  const dialog2 = toDialog(dialogRoot);
  if (!dialog2) {
    return;
  }
  const sourceElement = dialogRoot && dialogRoot instanceof HTMLElement && dialogRoot !== dialog2 ? dialogRoot : document.activeElement;
  if (sourceElement) {
    lastFocusByDialog.set(dialog2, sourceElement);
  }
  setOpen(dialog2, true);
}
function close(dialogRoot) {
  setOpen(toDialog(dialogRoot), false);
}
function handleDocumentKeydown(event) {
  if (event.key === "Escape") {
    const dialog2 = document.querySelector(ACTIVE_DIALOG_SELECTOR);
    if (dialog2) {
      close(dialog2);
      event.preventDefault();
    }
    return;
  }
  if (event.key !== "Tab") {
    return;
  }
  const activeDialog = document.querySelector(ACTIVE_DIALOG_SELECTOR);
  if (!activeDialog || !activeDialog.contains(event.target)) {
    return;
  }
  const items = findFocusable(activeDialog);
  if (items.length === 0) {
    event.preventDefault();
    return;
  }
  const first = items[0];
  const last = items[items.length - 1];
  if (!first || !last) {
    event.preventDefault();
    return;
  }
  const current = document.activeElement;
  if (event.shiftKey) {
    if (current === first || current === activeDialog) {
      last.focus();
      event.preventDefault();
    }
  } else if (current === last || current === activeDialog) {
    first.focus();
    event.preventDefault();
  }
}
function handleDocumentClick(event) {
  const target = event.target;
  if (!target) {
    return;
  }
  const openButton = target.closest("[data-ui8kit-dialog-open]");
  if (openButton) {
    open(openButton);
    event.preventDefault();
    return;
  }
  if (target.closest("[data-ui8kit-dialog-close]")) {
    close(target.closest(DIALOG_SELECTOR));
    event.preventDefault();
    return;
  }
  if (target.closest("[data-ui8kit-dialog-overlay]")) {
    close(target.closest(DIALOG_SELECTOR));
  }
}
var dialog = {
  name: "dialog",
  init(root = document) {
    const scope = root;
    const dialogs = scope.querySelectorAll(DIALOG_SELECTOR);
    for (const node of dialogs) {
      setOpen(node, node.getAttribute(OPEN_ATTR) === OPEN_VALUE);
    }
    document.addEventListener("click", handleDocumentClick);
    document.addEventListener("keydown", handleDocumentKeydown);
    return () => {
      document.removeEventListener("click", handleDocumentClick);
      document.removeEventListener("keydown", handleDocumentKeydown);
    };
  },
  open,
  close
};

// src/entries/dialog.ts
registerPattern(dialog);
var init2 = dialog.init;
var open2 = dialog.open;
var close2 = dialog.close;

exports.close = close2;
exports.init = init2;
exports.open = open2;
