// @ui8kit/aria/dist/disclosure.cjs
'use strict';

// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/disclosure/index.ts
function setState(root, open) {
  const trigger = root.querySelector("[data-disclosure-trigger]");
  const content = root.querySelector("[data-disclosure-content]");
  if (!trigger || !content) {
    return;
  }
  trigger.setAttribute("aria-expanded", String(open));
  if (open) {
    content.removeAttribute("hidden");
  } else {
    content.setAttribute("hidden", "hidden");
  }
}
var disclosure = {
  name: "disclosure",
  init(root = document) {
    const roots = root.querySelectorAll('[data-ui8kit="disclosure"]');
    for (const disclosureRoot of roots) {
      const trigger = disclosureRoot.querySelector("[data-disclosure-trigger]");
      const content = disclosureRoot.querySelector("[data-disclosure-content]");
      if (!trigger || !content || disclosureRoot.dataset.ui8kitBound) continue;
      const initial = trigger.getAttribute("aria-expanded") === "true";
      setState(disclosureRoot, initial);
      trigger.addEventListener("click", () => {
        const next = trigger.getAttribute("aria-expanded") !== "true";
        setState(disclosureRoot, next);
      });
      trigger.addEventListener("keydown", (event) => {
        if (event.key === "Enter" || event.key === " ") {
          event.preventDefault();
          trigger.click();
        }
      });
      disclosureRoot.dataset.ui8kitBound = "1";
    }
    return () => {
    };
  }
};

// src/entries/disclosure.ts
registerPattern(disclosure);
var init2 = disclosure.init;

exports.init = init2;
