// @ui8kit/aria/dist/listbox.mjs
// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/listbox/index.ts
function setSelected(root, target) {
  const options = root.querySelectorAll('[role="option"]');
  for (const option of options) {
    option.classList.remove("ui-listbox-option-active");
    option.setAttribute("aria-selected", "false");
    option.setAttribute("tabindex", "-1");
  }
  target.classList.add("ui-listbox-option-active");
  target.setAttribute("aria-selected", "true");
  target.setAttribute("tabindex", "0");
  target.focus();
}
var listbox = {
  name: "listbox",
  init(root = document) {
    const lists = root.querySelectorAll('[data-ui8kit="listbox"]');
    for (const list of lists) {
      const options = Array.from(list.querySelectorAll('[role="option"]'));
      if (!options.length || list.dataset.ui8kitBound) continue;
      for (let i = 0; i < options.length; i++) {
        options[i].tabIndex = i === 0 ? 0 : -1;
        options[i].addEventListener("click", () => {
          setSelected(list, options[i]);
        });
      }
      list.addEventListener("keydown", (event) => {
        if (event.key === "Home") {
          setSelected(list, options[0]);
          event.preventDefault();
        }
        if (event.key === "End") {
          setSelected(list, options[options.length - 1]);
          event.preventDefault();
        }
        if (event.key === "ArrowDown") {
          const current = options.indexOf(document.activeElement);
          const next = (current + 1) % options.length;
          setSelected(list, options[next]);
          event.preventDefault();
        }
        if (event.key === "ArrowUp") {
          const current = options.indexOf(document.activeElement);
          const prev = (current - 1 + options.length) % options.length;
          setSelected(list, options[prev]);
          event.preventDefault();
        }
      });
      list.dataset.ui8kitBound = "1";
    }
    return () => {
    };
  }
};

// src/entries/listbox.ts
registerPattern(listbox);
var init2 = listbox.init;

export { init2 as init };
