// @ui8kit/aria/dist/menu.cjs
'use strict';

// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/menu/index.ts
function makeItemsFocusable(items) {
  for (const [index, item] of items.entries()) {
    item.tabIndex = index === 0 ? 0 : -1;
  }
}
function moveFocus(items, current, delta) {
  const next = (current + delta + items.length) % items.length;
  items[current].tabIndex = -1;
  items[next].tabIndex = 0;
  items[next].focus();
  return next;
}
var menu = {
  name: "menu",
  init(root = document) {
    const menus = root.querySelectorAll('[data-ui8kit="menu"]');
    for (const menuRoot of menus) {
      const items = Array.from(menuRoot.querySelectorAll("[data-menu-item]"));
      if (!items.length || menuRoot.dataset.ui8kitBound) continue;
      makeItemsFocusable(items);
      let current = 0;
      for (const item of items) {
        item.addEventListener("keydown", (event) => {
          if (event.key === "ArrowDown") {
            current = moveFocus(items, current, 1);
            event.preventDefault();
          }
          if (event.key === "ArrowUp") {
            current = moveFocus(items, current, -1);
            event.preventDefault();
          }
          if (event.key === "Home") {
            current = moveFocus(items, current, -current);
            event.preventDefault();
          }
          if (event.key === "End") {
            current = moveFocus(items, current, items.length - 1 - current);
            event.preventDefault();
          }
          if (event.key === "Escape") {
            menuRoot.setAttribute("hidden", "hidden");
            const button = document.querySelector(`[data-menubutton-target="${menuRoot.id}"]`);
            button == null ? void 0 : button.focus();
          }
        });
        item.addEventListener("click", () => {
          item.focus();
        });
      }
      menuRoot.dataset.ui8kitBound = "1";
    }
    return () => {
    };
  }
};

// src/entries/menu.ts
registerPattern(menu);
var init2 = menu.init;

exports.init = init2;
