// @ui8kit/aria/dist/menubutton.mjs
// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/menubutton/index.ts
function getMenu(button) {
  const id = button.dataset.menubuttonTarget || button.getAttribute("aria-controls");
  if (!id) return null;
  const menu = document.getElementById(id);
  return menu && menu.matches('[data-ui8kit="menu"]') ? menu : null;
}
function closeAllMenus() {
  const menus = document.querySelectorAll('[data-ui8kit="menu"]');
  for (const menu of menus) {
    menu.setAttribute("hidden", "hidden");
    menu.dataset.state = "closed";
  }
  const buttons = document.querySelectorAll('[data-ui8kit="menubutton"]');
  for (const btn of buttons) {
    btn.setAttribute("aria-expanded", "false");
  }
}
function toggle(button) {
  const menu = getMenu(button);
  if (!menu) return;
  const isOpen = menu.dataset.state === "open";
  if (isOpen) {
    menu.setAttribute("hidden", "hidden");
    menu.dataset.state = "closed";
    button.setAttribute("aria-expanded", "false");
  } else {
    closeAllMenus();
    menu.removeAttribute("hidden");
    menu.dataset.state = "open";
    button.setAttribute("aria-expanded", "true");
    const first = menu.querySelector("[data-menu-item]");
    first == null ? void 0 : first.focus();
  }
}
function handleDocumentClick(event) {
  const target = event.target;
  if (!target) return;
  if (!target.closest('[data-ui8kit="menubutton"], [data-ui8kit="menu"]')) {
    closeAllMenus();
  }
}
var menubutton = {
  name: "menubutton",
  init(root = document) {
    const scope = root;
    const buttons = scope.querySelectorAll('[data-ui8kit="menubutton"]');
    const cleanups = [];
    for (const button of buttons) {
      if (button.dataset.ui8kitBound) continue;
      button.setAttribute("aria-expanded", "false");
      const onClick = (event) => {
        event.preventDefault();
        toggle(button);
      };
      const onKeydown = (event) => {
        if (event.key === "ArrowDown" || event.key === "Enter" || event.key === " ") {
          event.preventDefault();
          toggle(button);
        }
      };
      button.addEventListener("click", onClick);
      button.addEventListener("keydown", onKeydown);
      button.dataset.ui8kitBound = "1";
      cleanups.push(() => {
        button.removeEventListener("click", onClick);
        button.removeEventListener("keydown", onKeydown);
      });
    }
    document.addEventListener("click", handleDocumentClick);
    return () => {
      document.removeEventListener("click", handleDocumentClick);
      for (const fn of cleanups) fn();
    };
  },
  closeAll: closeAllMenus
};

// src/entries/menubutton.ts
registerPattern(menubutton);
var init2 = menubutton.init;
var close = menubutton.closeAll;

export { close, init2 as init };
