// @ui8kit/aria/dist/switch.cjs
'use strict';

// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/switch/index.ts
function toggle(control) {
  const checked = control.getAttribute("aria-checked") === "true";
  const next = !checked;
  control.setAttribute("aria-checked", String(next));
  control.setAttribute("data-state", next ? "checked" : "unchecked");
  control.dataset.value = String(next);
}
var switchPattern = {
  name: "switch",
  init(root = document) {
    var _a;
    const switches = root.querySelectorAll('[data-ui8kit="switch"], [role="switch"]');
    for (const item of switches) {
      if (item.dataset.ui8kitBound) continue;
      item.dataset.ui8kitBound = "1";
      const control = item.querySelector("[data-switch-control]") || item;
      control.setAttribute("role", "switch");
      control.setAttribute("tabindex", (_a = control.getAttribute("tabindex")) != null ? _a : "0");
      if (!control.hasAttribute("aria-checked")) {
        control.setAttribute("aria-checked", "false");
      }
      const handler = (event) => {
        event.preventDefault();
        toggle(control);
      };
      control.addEventListener("click", handler);
      control.addEventListener("keydown", (event) => {
        if (event.key === " " || event.key === "Enter") {
          handler(event);
        }
      });
    }
    return () => {
    };
  }
};

// src/entries/switch.ts
registerPattern(switchPattern);
var init2 = switchPattern.init;

exports.init = init2;
