// @ui8kit/aria/dist/tabs.cjs
'use strict';

// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/tabs/index.ts
function activate(root, value, useFocus) {
  const triggers = root.querySelectorAll("[data-tabs-trigger]");
  const panels = root.querySelectorAll("[data-tabs-panel]");
  for (const trigger of triggers) {
    const selected = trigger.getAttribute("data-tabs-value") === value;
    trigger.setAttribute("aria-selected", selected ? "true" : "false");
    trigger.tabIndex = selected ? 0 : -1;
    if (selected && useFocus) {
      trigger.focus();
    }
  }
  for (const panel of panels) {
    panel.hidden = panel.getAttribute("data-tabs-value") !== value;
  }
}
function defaultValue(root) {
  var _a;
  const rootValue = root.getAttribute("data-tabs-value");
  if (rootValue) return rootValue;
  const selected = root.querySelector('[data-tabs-trigger][aria-selected="true"]');
  if (selected == null ? void 0 : selected.getAttribute("data-tabs-value")) {
    return selected.getAttribute("data-tabs-value");
  }
  return ((_a = root.querySelector("[data-tabs-trigger]")) == null ? void 0 : _a.getAttribute("data-tabs-value")) || "";
}
function onKeydown(root, event) {
  const triggers = Array.from(root.querySelectorAll("[data-tabs-trigger]"));
  if (!triggers.length) return;
  const trigger = event.target.closest("[data-tabs-trigger]");
  if (!trigger) return;
  const index = triggers.indexOf(trigger);
  if (index < 0) return;
  if (event.key === "ArrowRight" || event.key === "ArrowDown") {
    const next = triggers[(index + 1) % triggers.length];
    event.preventDefault();
    activate(root, next.getAttribute("data-tabs-value") || "", true);
  } else if (event.key === "ArrowLeft" || event.key === "ArrowUp") {
    const prev = triggers[(index - 1 + triggers.length) % triggers.length];
    event.preventDefault();
    activate(root, prev.getAttribute("data-tabs-value") || "", true);
  }
}
var tabs = {
  name: "tabs",
  init(root = document) {
    const roots = root.querySelectorAll('[data-ui8kit="tabs"]');
    for (const tabsRoot of roots) {
      if (tabsRoot.dataset.ui8kitBound) continue;
      tabsRoot.dataset.ui8kitBound = "1";
      const value = defaultValue(tabsRoot);
      if (value) {
        activate(tabsRoot, value, false);
      }
      tabsRoot.addEventListener("click", (event) => {
        const trigger = event.target.closest("[data-tabs-trigger]");
        if (!trigger) return;
        const value2 = trigger.getAttribute("data-tabs-value");
        if (!value2) return;
        event.preventDefault();
        activate(tabsRoot, value2, false);
      });
      tabsRoot.addEventListener("keydown", (event) => {
        onKeydown(tabsRoot, event);
      });
    }
    return () => {
    };
  }
};

// src/entries/tabs.ts
registerPattern(tabs);
var init2 = tabs.init;

exports.init = init2;
