// @ui8kit/aria/dist/tooltip.mjs
// src/core/utils.ts
function ready(fn) {
  if (typeof document === "undefined") {
    fn();
    return;
  }
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", fn, { once: true });
    return;
  }
  fn();
}
function byAttr(name, root = document) {
  return root.querySelectorAll(`[data-${name}]`);
}
function isBrowser() {
  return typeof document !== "undefined" && typeof window !== "undefined";
}

// src/core/env.ts
var AUTO_INIT_KEY = "__UI8KIT_ARIA_AUTO_INIT__";
function isAutoInitEnabled() {
  return globalThis[AUTO_INIT_KEY] !== false;
}

// src/core/registry.ts
var NAMESPACE_KEY = "ui8kit";
var state = {
  patterns: /* @__PURE__ */ new Map(),
  scopes: /* @__PURE__ */ new Map()
};
var cachedNamespace = null;
var coreHelpers = { ready, byAttr };
function rootKey(root) {
  return root != null ? root : isBrowser() ? document : {};
}
function buildNamespace() {
  const existing = globalThis[NAMESPACE_KEY];
  if (existing) {
    return existing;
  }
  const ns = {
    core: coreHelpers,
    ready,
    byAttr,
    register: registerPattern,
    init,
    initPattern
  };
  if (isBrowser()) {
    globalThis[NAMESPACE_KEY] = ns;
  }
  return ns;
}
function getNamespace() {
  if (!cachedNamespace) {
    cachedNamespace = buildNamespace();
  }
  return cachedNamespace;
}
function registerPattern(pattern) {
  if (state.patterns.has(pattern.name)) {
    return;
  }
  state.patterns.set(pattern.name, pattern);
  getNamespace()[pattern.name] = pattern;
  if (isAutoInitEnabled() && isBrowser()) {
    ready(() => {
      initPattern(pattern.name);
    });
  }
}
function initPattern(name, root) {
  var _a;
  const pattern = state.patterns.get(name);
  if (!pattern) {
    return;
  }
  if (!isBrowser() && !root) {
    return;
  }
  const scope = rootKey(root);
  const scopeMap = (_a = state.scopes.get(name)) != null ? _a : /* @__PURE__ */ new WeakMap();
  state.scopes.set(name, scopeMap);
  if (scopeMap.has(scope)) {
    return;
  }
  const result = pattern.init(scope);
  scopeMap.set(scope, typeof result === "function" ? result : noop);
}
function init(root) {
  for (const pattern of state.patterns.values()) {
    initPattern(pattern.name, root);
  }
}
function noop() {
}

// src/patterns/tooltip/index.ts
function show(root) {
  const content = root.querySelector('[role="tooltip"]');
  if (!content) {
    return;
  }
  content.removeAttribute("hidden");
  root.setAttribute("data-state", "open");
  content.setAttribute("aria-hidden", "false");
}
function hide(root) {
  const content = root.querySelector('[role="tooltip"]');
  if (!content) {
    return;
  }
  content.setAttribute("hidden", "hidden");
  root.setAttribute("data-state", "closed");
  content.setAttribute("aria-hidden", "true");
}
var tooltip = {
  name: "tooltip",
  init(root = document) {
    const tooltips = root.querySelectorAll('[data-ui8kit="tooltip"]');
    for (const tooltipRoot of tooltips) {
      if (tooltipRoot.dataset.ui8kitBound) continue;
      tooltipRoot.addEventListener("mouseenter", () => show(tooltipRoot));
      tooltipRoot.addEventListener("focusin", () => show(tooltipRoot));
      tooltipRoot.addEventListener("mouseleave", () => hide(tooltipRoot));
      tooltipRoot.addEventListener("focusout", () => hide(tooltipRoot));
      tooltipRoot.dataset.ui8kitBound = "1";
    }
    return () => {
    };
  }
};

// src/entries/tooltip.ts
registerPattern(tooltip);
var init2 = tooltip.init;

export { init2 as init };
